apps/finevent/finevent.star

```diff
@@ -0,0 +1,580 @@
+"""
+Applet: Finevent
+Summary: Upcoming financial events
+Description: Displays the daily economic or earnings calendar.
+Author: Rob Kimball
+"""
+
+load("http.star", "http")
+load("time.star", "time")
+load("cache.star", "cache")
+load("render.star", "render")
+load("schema.star", "schema")
+load("encoding/json.star", "json")
+load("encoding/base64.star", "base64")
+
+BASE_URL = "https://api.tradingeconomics.com"
+AUTH = "guest:guest"
```

Comment:
Hey Mark - your concern is well received and completely fair. My understanding is that the `guest:guest` credential (used throughout their docs) is a preview of their API. The data available is very limited, but I found it to be adequate for this specific use case. I've been hitting it every 3 minutes for the past 4 days (importance 2 & 3, so 960 requests/day) using `pixlet render` (so no cache is available) and haven't noticed any interruptions or limitations - in production you'll notice the events are cached until the next release in the queue, so this is an overestimation of what will actually be needed.

Looking at the past month of releases, there were 77 times throughout the day where new data was released (including all countries at all levels of importance), the most common being on each hour from 3-8 AM EST, and 8:30 AM or 10:00 AM for most US data. On a daily basis, there were at most 48 unique release times (on 4/7, lots of BOT releases coming this week following month end), but usually there are ~26 (median), and on the weekends there are usually only 3-4.

![image](https://user-images.githubusercontent.com/7003930/161628331-42aa5b96-0cbb-47ee-bb98-bb9066563d45.png)

As it's currently configured, there are 6 request URLs: Global or US-only, and one for each of the 3 levels of importance:

- https://api.tradingeconomics.com/calendar/country/all?c=guest:guest&f=json&importance=3
- https://api.tradingeconomics.com/calendar/country/all?c=guest:guest&f=json&importance=2
- https://api.tradingeconomics.com/calendar/country/all?c=guest:guest&f=json&importance=1
- https://api.tradingeconomics.com/calendar/country/united%20states?c=guest:guest&f=json&importance=3
- https://api.tradingeconomics.com/calendar/country/united%20states?c=guest:guest&f=json&importance=2
- https://api.tradingeconomics.com/calendar/country/united%20states?c=guest:guest&f=json&importance=1

In practice we won't need to touch all 6 at every release time, but as a very conservative estimate: `48 * 6 = 288`, which is comfortably under the number of times I've been requesting data during testing.

I just tried stress-testing this by sending 3 requests per second and was finally able to get a 403 back after 2 minutes, so I suspect as long as we're caching globally across all devices we will be fine. This isn't the only economic events API out there and I'm open to trying others, but this was certainly the easiest/most complete one I found for global coverage, and could open the door for other applets down the line should you and the team decide a premium license is justifiable.

Reply:
As a robustness check, I think installing this on your side to a device or two and monitoring it for the next week would be prudent, especially since this Thursday will be a high volume day. I'll keep the render/push loop running on my local as well to see if anything breaks.

Reply:
Wow! Thanks for such a detailed response! Ok, let's merge this and give it a go. The biggest risk is this becomes too popular and we need to start paying for a license or remove it because the license is too expensive.